Implement custom ruby enumerable methods

Teshager21 opened the pull request:
- use git-flow 
- implement my own each for ruby collections(range, array, hashes) using yield
- built the other methods using the my_each as a base
- test the inject method using multiply_els 
-test all of the methods using the examples on the ruby enumerable documentation

Teshager21 committed:
test the linters

Teshager21 committed:
add my_each method for array input

Teshager21 committed:
fix linter errors

Teshager21 committed:
add my_each_with_index method

Teshager21 committed:
add my_select method

Teshager21 committed:
add my_ll method

Teshager21 committed:
add #my_any

Teshager21 committed:
add #my-none?

Teshager21 committed:
add #my_count

Teshager21 committed:
add #my_map

Teshager21 committed:
add #myinject

Teshager21 committed:
use mulitply_els ot test my_inject

Teshager21 committed:
include ranges in #my_each

Teshager21 committed:
include hashe in #my_each method

Teshager21 committed:
remove commented code

Teshager21 committed:
add screenshot to the readme

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+        yield keys[i], values[i]
+      end
+    end
+  end
```
`my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods)) ``
- After calling the given block once for each element in self, returns the array itself
```ruby
[1, 2, 3].my_each {|num| num}  #should return [1, 2, 3]
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+    else
+      to_enum
+    end
+  end
```
`my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
-  Calls the given block once for each element in self, the block should accept two values value, index. The first one is value and the second one is the index. But here you have altered the sequence.
- After calling the given block once for each element in self, returns the array itself
```ruby
[1, 2, 3].my_each_with_index {|num| num}  #should return [1, 2, 3]
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+      temp << i if yield i
+    end
+    temp
+  end
```
`my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- Should return an enumerator if no block given.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+      end
+    end
+    false
+  end
```
`my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true, that it is doing
[1, 2, nil].my_any? #should return true, provides correct output
[false, false, nil].my_any? #should return false, it failed to test it
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+    end
+
+    mapped
+  end
```
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- returns an Enumerator if no block is given
![image](https://user-images.githubusercontent.com/38625956/83134736-99c63180-a106-11ea-9c35-3dff0fb877e6.png)
- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
my_proc = Proc.new { |num| num > 10 }
[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+    mapped
+  end
+
+  def my_inject(arg = nil, symb = nil)
+    temp_self = if is_a? Range
+                  to_a
+                else
+                  self
+                end
+
+    memo = if arg.nil? or arg.is_a? Symbol
+             first
+           else
+             arg
+           end
+    symb = arg if arg.is_a? Symbol
+    unless symb.nil?
+      mymethod = symb.to_s
+      for item in 1...to_a.length
+
+        memo = eval "#{memo}#{mymethod}#{temp_self[item]}"
+      end
+    end
+    if block_given?
+      for item in 1...length
+        memo = yield memo, self[item]
+      end
+    end
+    memo
+  end
```
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
(1..3).my_inject(1) { |memo, num| memo + num } #should return 7
%w[dog door rod blade].my_inject {|memo, word| memo.length > word.length ? memo : word}  #should return 'blade'
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(1, :+)  #should return 7
(1..3).my_inject(1, :+)  #should return 7
(5..10).my_inject(2, :*)  #should return 302400
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,190 @@
+    memo
+  end
+
+  def multiply_els
+    my_inject(:*)
+  end
```
multiply_els is not an enumerable method. So it should be placed outside of the module. Then it should appear something like this.
```ruby
def multiply_els(arr)
  arr.my_inject(2, :*)
end
```

shshamim63 reviewed (changes requested):
## Status :traffic_light: Required Changes :tractor: 
___
Awesome initiative. Most of your methods are working properly, repo contains a descriptive readme. But there are some issues that are blocking me to approve this now(The test I have attached as the example are also derived from the documentation, ruby-doc, so If you think that some tests are totally contradictory don't hesitate to knock me, I would love to describe them). Let's get to those issues.
___
### :spades: Mandatory Changes :spades: 
___
The following methods contain edge cases and please check out the comment added to their blocks.
- my_each
- my_each_with_index
- my_select
- my_any?
- my_map
- my_inject
- multiply_els
___
##### Please resolve the issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

Teshager21 committed:
worke on edge cases

Teshager21 committed:
remove test expressions

Teshager21 commented:
**I have implemeted all the changes requested by the TSE**

- I have accounted for all the edge cases listed out by the TSE
- Moved the multiply_els out of the module
- Accounted for empty block passes
- Made my-map prioritize the proc when both proc and block are given

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,210 @@
+    else
+      to_enum
+    end
+  end
```
https://github.com/Teshager21/ruby-enumerables/pull/1#discussion_r431734409 I meant that you the previous data you have passed in your block `yield i, self[i]`. But the sequence should be `yield self[i], i` So the ultimate formation of this section should be
```ruby
def my_each_with_index
  return to_enum unless block_given?
  for i in (0...length)
    yield self[i], i
  end
  self
end
```
So please replace your my_each_with_index with the above one.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,210 @@
+    end
+
+    mapped
+  end
```
Congrats you have covered the edge case mentioned here https://github.com/Teshager21/ruby-enumerables/pull/1#discussion_r431738511 But it is not covering the simple case now, For an example,
```ruby
p [1, 2, 3].my_map {|num| num < 10 } #should return [true, true, true]
```
But I am getting an error. Because  my_map expects an argument. So Let's fix it. Replace the function declaration from `def my_map(arg)` to `def my_map(arg = nil)` And it will resolve the problem.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,210 @@
+      end
+    end
+    memo
+  end
```
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
___
You have covered the second part of this comment https://github.com/Teshager21/ruby-enumerables/pull/1#discussion_r431752606 But the following part is still uncovered.
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
(1..3).my_inject(1) { |memo, num| memo + num } #should return 7
%w[dog door rod blade].my_inject {|memo, word| memo.length > word.length ? memo : word}  #should return 'blade'
```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,210 @@
+
+def multiply_els
+  my_inject(:*)
+end
```
Are you sure this method is callable with an array? I already provided you the example solution here https://github.com/Teshager21/ruby-enumerables/pull/1#discussion_r431754626. May I know you did not like it through slack. Eagerly waiting for the cause. Please replace multiply_els with the solution provided in the previous comment.

shshamim63 reviewed (changes requested):
## Status :red_circle: Required Changes 
___
Hi @Teshager21 
Almost covered all the changes, but there are some corner cases that are still in there unresolved. Please read the comment attached in the block of codes carefully and if anything seems unclear feel free to reach me.
___
### :spades: Mandatory Changes :spades: 
___
- Replace your my_each_with_index with the sample that I have given.
- Follow the instruction attached to my_map block and fix it.
- Your inject is not fully functional. Are you sure the following sample tests are providing an exact result? Cause I tried but was getting wrong answers.
```ruby
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
(1..3).my_inject(1) { |memo, num| memo + num } #should return 7
%w[dog door rod blade].my_inject {|memo, word| memo.length > word.length ? memo : word}  #should return 'blade'
```
- Check out the comment in the multiply_els section and fix it.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

Teshager21 committed:
test #myinject for different inputs

Teshager21 committed:
accomodate empty arg input for #my_map

Teshager21 committed:
clear out the test cases

Teshager21 commented:
**I have implemented all the changes requested by the TSE**

- I have accounted for all the edge cases listed out by the TSE
- Moved the multiply_els out of the module and adjusted it
- Accounted for empty block passes
- Made my-map prioritize the proc when both proc and block are given
-Modified to account for the remaining cases

Forison reviewed (approved):
# Status:: Approved✔️
## Good on implementing the issues raised earlier, all the methods work as required, go ahead and merge the branch and proceed to your next task😎
#### Happy coding🤗
Addo
